### PR TITLE
Config as Data Overview

### PR DESCRIPTION
**Pull request by martinmaly**
The first of several documents describing the foundation and
implementation of Package Orchestration.

### CODE REVIEW

File: docs/design-docs/06-config-as-data.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  structures and other variations within the model
+* decouples abstractions of configuration from collections of configuration data
+* represents abstractions of configuration generators as data with schemas, like
+  other configuration data
+* finds, filters / queries / selects, and/or validates configuration data that
+  can be operated on by given code (functions)
+* finds and/or filters / queries / selects code (functions) that can operate on
+  resource types contained within a body of configuration data
+* separates configuration data transformations from actuation, enabling
+  actuation by performing bulk changes to the configuration data; *actuation*,
```
Comment (bgrant0607) [inline docs/design-docs/06-config-as-data.md:49]: The partial definition of actuation is a bit hard to parse with this wording. Is "actuation" intended to be a separate bullet? Or "with actuation including..."?

Really this implies that reconciliation with the live state itself needs to be driven declaratively.

File: docs/design-docs/06-config-as-data.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+  including sequencing, orchestration, and dependent data resolution occurs in
+  the actuation layers
+* transformations, particular value propagation, are preferable to wholesale
+  configuration generation except when the expansion is dramatic (say, >10x)
+* transformation input generation should usually be decoupled from propagation
+* deployment context inputs should be taken from well defined “provider context”
+  objects identifiers and references should be declarative
```
Comment (bgrant0607) [inline docs/design-docs/06-config-as-data.md:56]: Identifiers should start a new bullet

File: docs/design-docs/06-config-as-data.md [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+[Kubernetes Resource Model](https://github.com/kubernetes/design-proposals-archive/blob/main/architecture/resource-management.md)
+(KRM).
+
+**Note**: Even though KRM is not a requirement of Config as Data (just like
+Python or Go templates or Jinja are not specifically requirements for
+[IaC](https://en.wikipedia.org/wiki/Infrastructure_as_code)), the choice of
+another foundational config reporesentation format would necessitate
+implementing adapters for all types of infrastructure and applications
+configured, including Kubernetes, CRDs, GCP resources and more. Likewise, choice
+of another configuration format owuld require redesign of a number of the
```
Comment (bgrant0607) [inline docs/design-docs/06-config-as-data.md:75]: Typo: s/owuld/would/

File: docs/design-docs/06-config-as-data.md [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  constraint-based policy enforcement, and disaster recovery
+* bulk changes to configuration data in their sources of truth
+* injection of configuration to address horizontal concerns
+* merging of multiple sources of truth
+* state export to reusable blueprints without manual templatization
+* cooperative editing of configuration by humans and automation, such as for
+  security remediation (which is usually implemented against live-state APIs)
+* reusability of configuration transformation code across multiple bodies of
+  configuration data containing the same resource types
+* amortizing the effort of writing, testing, documenting the code
```
Comment (bgrant0607) [inline docs/design-docs/06-config-as-data.md:141]: This should be a sub-bullet, or incorporated into the bullet above

File: docs/design-docs/06-config-as-data.md [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+* maintaining variants of configuration blueprints without one-size-fits-all
+  full struct-constructor-style parameterization and without manually
+  constructing and maintaining patches
+* drift detection and remediation for most of the desired state via continuous
+  reconciliation using apply and/or for specific attributes via targeted
+  mutation of the sources of truth
+
+## Related Articles
+
+For more information about Configuration as Data and Kubernetes Resource Model,
```
Comment (bgrant0607) [inline docs/design-docs/06-config-as-data.md:158]: There's also https://kpt.dev/guides/rationale

File: docs/design-docs/06-config-as-data.md [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+  other configuration data
+* finds, filters / queries / selects, and/or validates configuration data that
+  can be operated on by given code (functions)
+* finds and/or filters / queries / selects code (functions) that can operate on
+  resource types contained within a body of configuration data
+* separates configuration data transformations from actuation, enabling
+  actuation by performing bulk changes to the configuration data; *actuation*,
+  including sequencing, orchestration, and dependent data resolution occurs in
+  the actuation layers
+* transformations, particular value propagation, are preferable to wholesale
```
Comment (bgrant0607) [inline docs/design-docs/06-config-as-data.md:52]: particularly

**Review by bgrant0607 — COMMENTED** [review]
Thanks! Just some minor editing issues.

File: docs/design-docs/06-config-as-data.md (hunk 2 above)
Comment (martinmaly) [reply docs/design-docs/06-config-as-data.md:56]: thanks!

File: docs/design-docs/06-config-as-data.md (hunk 3 above)
Comment (martinmaly) [reply docs/design-docs/06-config-as-data.md:75]: thanks!

File: docs/design-docs/06-config-as-data.md (hunk 6 above)
Comment (martinmaly) [reply docs/design-docs/06-config-as-data.md:52]: fixed; thanks

File: docs/design-docs/06-config-as-data.md (hunk 5 above)
Comment (martinmaly) [reply docs/design-docs/06-config-as-data.md:158]: added! Thanks!

File: docs/design-docs/06-config-as-data.md (hunk 6 above)
Comment (martinmaly) [reply docs/design-docs/06-config-as-data.md:52]: fixed

File: docs/design-docs/06-config-as-data.md (hunk 4 above)
Comment (martinmaly) [reply docs/design-docs/06-config-as-data.md:141]: fixed; thanks!

File: docs/design-docs/06-config-as-data.md (hunk 1 above)
Comment (martinmaly) [reply docs/design-docs/06-config-as-data.md:49]: Rewrote; LMK if this is better.

**Review by bgrant0607 — APPROVED** [review]
